nifi/nifi-nar-bundles/nifi-flume-bundle/nifi-flume-processors/src/main/java/org/apache/nifi/processors/flume/FlumeSourceProcessor.java

```diff
@@ -0,0 +1,191 @@
+public class FlumeSourceProcessor extends AbstractFlumeProcessor {
+
+    private Source source;
+    private SourceRunner runner;
+    private MemoryChannel channel;
+
+    public static final PropertyDescriptor SOURCE_TYPE = new PropertyDescriptor.Builder()
+            .name("Source Type")
+            .description("The fully-qualified name of the Source class")
+            .required(true)
+            .addValidator(createSourceValidator())
+            .build();
+    public static final PropertyDescriptor AGENT_NAME = new PropertyDescriptor.Builder()
+            .name("Agent Name")
+            .description("The name of the agent used in the Flume source configuration")
+            .required(true)
+            .defaultValue("tier1")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor SOURCE_NAME = new PropertyDescriptor.Builder()
+            .name("Source Name")
+            .description("The name of the source used in the Flume source configuration")
+            .required(true)
+            .defaultValue("src-1")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor FLUME_CONFIG = new PropertyDescriptor.Builder()
+            .name("Flume Configuration")
+            .description("The Flume configuration for the source copied from the flume.properties file")
+            .required(true)
+            .defaultValue("")
+            .addValidator(Validator.VALID)
+            .build();
+
+    public static final Relationship SUCCESS = new Relationship.Builder().name("success").build();
+
+    private List<PropertyDescriptor> descriptors;
+    private Set<Relationship> relationships;
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        this.descriptors = ImmutableList.of(SOURCE_TYPE, AGENT_NAME, SOURCE_NAME, FLUME_CONFIG);
+        this.relationships = ImmutableSet.of(SUCCESS);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @OnScheduled
+    public void onScheduled(final SchedulingContext context) {
+        try {
+            source = SOURCE_FACTORY.create(
+                    context.getProperty(SOURCE_NAME).getValue(),
+                    context.getProperty(SOURCE_TYPE).getValue());
+
+            String flumeConfig = context.getProperty(FLUME_CONFIG).getValue();
+            String agentName = context.getProperty(AGENT_NAME).getValue();
+            String sourceName = context.getProperty(SOURCE_NAME).getValue();
+            Configurables.configure(source,
+                    getFlumeSourceContext(flumeConfig, agentName, sourceName));
+
+            if (source instanceof EventDrivenSource) {
```

Comment:
If I understand correctly, the memory channel is used to buffer incoming events between calls to onTrigger because there is no session with which the events would be sent. But this makes all event-driven sources unsafe because the events that arrive between onTrigger calls are stored in memory.

I think you instead need access to a ProcessSessionFactory. It looks like the Processor interface is passed a session factory when onTrigger is called, and [`AbstractProcessor`](https://github.com/apache/incubator-nifi/blob/develop/nifi/nifi-api/src/main/java/org/apache/nifi/processor/AbstractProcessor.java#L24) is what is giving you a session most of the time. I'm not sure if you can get one otherwise or if you can cache the instance, but I think this should be adding content directly to the repository or using a file channel (which can hopefully be avoided) rather than using a mem channel.